core/app/services/spree/add_item_to_cart.rb

```diff
@@ -0,0 +1,37 @@
+module Spree
+  class AddItemToCart
+    prepend Spree::ServiceModule::Base
+
+    def call(order:, variant:, quantity: nil, options: {})
+      run :add_to_line_item
+      run Spree::AfterAddOrRemove.new
+    end
+
+    private
+
+    def add_to_line_item(order:, variant:, quantity: nil, options: {})
+      options ||= {}
+      quantity ||= 1
+
+      line_item = Spree::FindLineItemByVariant.new.execute(order: order, variant: variant, options: options)
+
+      line_item_created = line_item.nil?
+      if line_item.nil?
+        opts = ::Spree::PermittedAttributes.line_item_attributes.each_with_object({}) do |attribute, result|
+          result[attribute] = options[attribute]
```

Comment:
I like the general idea but we'll have to refactor this since it's too heavy & hard to understand.